core/solver/batch_solver.ipp

```diff
@@ -0,0 +1,146 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+#define GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+
+
+#include <ginkgo/core/log/batch_convergence.hpp>
+#include <ginkgo/core/solver/batch_solver.hpp>
+
+
+#include "core/log/batch_logging.hpp"
+#include "core/matrix/batch_csr_kernels.hpp"
+#include "core/matrix/batch_dense_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace batch {
+
+
+GKO_REGISTER_OPERATION(pre_diag_scale_system, batch_csr::pre_diag_scale_system);
+GKO_REGISTER_OPERATION(vec_scale, batch_dense::batch_scale);
+
+
+}
+
+
+struct BatchInfo {
+    //gko::log::BatchLogData<ValueType> logdata;
+    void* logdata;
+};
+
+
+template <typename ConcreteSolver, typename PolymorphicBase>
+void EnableBatchSolver<ConcreteSolver, PolymorphicBase>::apply_impl(const BatchLinOp* b,
+                                          BatchLinOp* x) const
+{
+    using value_type = typename ConcreteSolver::value_type;
+    using Csr = matrix::BatchCsr<value_type>;
+    using Vector = matrix::BatchDense<value_type>;
+    using real_type = remove_complex<value_type>;
+
+    auto exec = this->get_executor();
+    auto dense_b = as<const Vector>(b);
+    auto dense_x = as<Vector>(x);
+    const bool to_scale =
+        this->get_left_scaling_vector() && this->get_right_scaling_vector();
+    const auto acsr = dynamic_cast<const Csr*>(system_matrix_.get());
+    const BatchLinOp* a_scaled{};
+    const Vector* b_scaled{};
+    auto a_scaled_smart = Csr::create(exec);
+    auto b_scaled_smart = Vector::create(exec);
+    if (to_scale && !acsr) {
+        GKO_NOT_SUPPORTED(system_matrix_);
+    }
+
+    // copies to scale
+    if (to_scale) {
+        a_scaled_smart->copy_from(acsr);
+        b_scaled_smart->copy_from(dense_b);
+        exec->run(batch::make_pre_diag_scale_system(
+            as<const Vector>(this->get_left_scaling_vector()),
+            as<const Vector>(this->get_right_scaling_vector()),
+            a_scaled_smart.get(), b_scaled_smart.get()));
+        a_scaled = a_scaled_smart.get();
+        b_scaled = b_scaled_smart.get();
+    } else {
+        a_scaled = system_matrix_.get();
+        b_scaled = dense_b;
+    }
+
+    // allocate logging arrays assuming uniform size batch
+    // GKO_ASSERT(dense_b->stores_equal_sizes());
+
+    const size_type num_rhs = dense_b->get_size().at(0)[1];
+    const size_type num_batches = dense_b->get_num_batch_entries();
+    batch_dim<> sizes(num_batches, dim<2>{1, num_rhs});
+
+    log::BatchLogData<value_type> concrete_logdata;
+    concrete_logdata.res_norms =
+        matrix::BatchDense<real_type>::create(this->get_executor(), sizes);
+    concrete_logdata.iter_counts.set_executor(this->get_executor());
+    concrete_logdata.iter_counts.resize_and_reset(num_rhs * num_batches);
```

Comment:
Is this actually for logging, or more like a `stopping_criterion` (it looks like the latter)?

Reply:
This is actually for logging information from inside the device kernel. It logs the final (implicit) residual values at the end of the solve, and the final iteration count. Since these values are otherwise never available on the host, we need some kind of logging for this.